File: code/ex914.java
class Solution {
    public boolean hasGroupsSizeX(int[] deck) {
        Map<Integer, Integer> cnt = new HashMap<>();
        for (int d : deck)    cnt.put(d, cnt.getOrDefault(d, 0) + 1);
        int GCD = -1;
        for (Map.Entry<Integer, Integer> entry : cnt.entrySet()) {
            if (GCD == -1)    GCD = entry.getValue();
            GCD = gcd(GCD, entry.getValue());
        }
        return GCD >= 2;
    }
    
    private int gcd(int a, int b) {
        if (b == 0)    return a;
        return gcd(b, a % b);
    }
}

File: code/ex914.cpp
class Solution {
public:
    bool hasGroupsSizeX(vector<int>& deck) {
        unordered_map<int, int> cnt;
        for (int d : deck)    cnt[d]++;
        int X = -1;
        for (auto& [key, value] : cnt) {
            if (X == -1)    X = value;
            else    X = GCD(X, value);
        }
        return X >= 2;
    }
    
private:
    int GCD(int a, int b) {
        if (b == 0)    return a;
        return GCD(b, a % b);
    }
};